### PR TITLE
[workflow] Test fault tolerance with storage

### LINKED ISSUE #17089: Strong fault tolerance testing

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

A thoroughly test to ensure we can resume the workflow correctly when all files after a certain time are lost.

## Related issue number

Closes #17089

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/storage/__init__.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 logger = logging.getLogger(__name__)
 
+DEBUG_PREFIX = "debug"
```
Comment (iycheng) [inline python/ray/experimental/workflow/storage/__init__.py:9]: it should be put into debug.py

File: python/ray/experimental/workflow/storage/__init__.py [hunk 2]
```diff
@@ -35,13 +37,18 @@ def create_storage(storage_url: str) -> Storage:
...
     elif parsed_url.scheme == "s3":
-        from ray.experimental.workflow.storage.s3 import (S3StorageImpl)
+        from ray.experimental.workflow.storage.s3 import S3StorageImpl
         bucket = parsed_url.netloc
         s3_path = parsed_url.path
         if not s3_path:
             raise ValueError(f"Invalid s3 path: {s3_path}")
         params = dict(parse.parse_qsl(parsed_url.query))
         return S3StorageImpl(bucket, s3_path, **params)
+    elif parsed_url.scheme == DEBUG_PREFIX:
```
Comment (iycheng) [inline python/ray/experimental/workflow/storage/__init__.py:47]: maybe just literal string here?

File: python/ray/experimental/workflow/storage/debug.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                with open(self._count, "x") as f:
+                    f.write("0")
+
+    async def put(self, key: str, data: Any, is_json: bool = False) -> None:
+        with FileLock(str(self._workflow_root_dir / ".lock")):
+            with open(self._count, "r") as f:
+                count = int(f.read())
+            k1 = self._log_dir / f"{count}.metadata.json"
+            k2 = self._log_dir / f"{count}.value"
+            await super().put(
```
Comment (iycheng) [inline python/ray/experimental/workflow/storage/debug.py:34]: should we just store it in this db's dir?

**Review by fishbone — COMMENTED** [review]
Mostly looks good to me.

File: python/ray/experimental/workflow/storage/debug.py (hunk 3 above)
Comment (suquark) [reply python/ray/experimental/workflow/storage/debug.py:34]: make sense

**Comment by iycheng** [comment]
nice job. thanks!

**Review by fishbone — APPROVED** [review]

**Pull request merged**